### PR TITLE
Improve detection of functions and abstract functions

### PR DESCRIPTION
**Pull request by mikehaertl**

### CODE REVIEW

File: ftplugin/php/phpfolding.vim [hunk 1]
```diff
@@ -181,10 +181,10 @@ function! s:PHPCustomFolds() " {{{
 	"       'g:searchEmptyLinesPostfixing'..
 
 	" Fold function with PhpDoc (function foo() {})
-	call s:PHPFoldPureBlock('\(^\|\s\)function\s\+[A-Za-z_][A-Za-z0-9_]*\s*(', s:FOLD_WITH_PHPDOC)
+	call s:PHPFoldPureBlock('^\(public\|private\|protected\|static\|\s\)*function\s\+[A-Za-z_][A-Za-z0-9_]*\s*(', s:FOLD_WITH_PHPDOC)
```
Comment (mikehaertl) [inline ftplugin/php/phpfolding.vim:184]: @rayburgemeestre This previously matched every `function` that has a space in front which is not good. It also included `abstract function` which broke the call below. Now it matches only lines with a proper function definition, with any of the visibility and static declarations in any order.

File: ftplugin/php/phpfolding.vim [hunk 2]
```diff
@@ -181,10 +181,10 @@ function! s:PHPCustomFolds() " {{{
...
 
 	" Fold abstract function with PhpDoc
-	call s:PHPFoldProperties('^\s*\(abstract\)\s\(\(private\)\|\(public\)\|\(protected\)\s\)*\(static\s\)*function\s\+[A-Za-z_][A-Za-z0-9_]*\s*(', ";", s:FOLD_WITH_PHPDOC, 1, 1)
+	call s:PHPFoldProperties('^\s*abstract\(public\|private\|protected\|static\|\s\)*function\s\+[A-Za-z_][A-Za-z0-9_]*\s*(', ";", s:FOLD_WITH_PHPDOC, 1, 1)
```
Comment (mikehaertl) [inline ftplugin/php/phpfolding.vim:187]: Here the brackets were not necessary and even misleading. Again the `static` and visibility keywords can appear now on any order (which is valid PHP syntax).

File: ftplugin/php/phpfolding.vim [hunk 3]
```diff
@@ -181,13 +181,13 @@ function! s:PHPCustomFolds() " {{{
...
 
-	" Fold class properties with PhpDoc (var $foo = NULL;)
-	call s:PHPFoldProperties('^\s*\(\(private\)\|\(public\)\|\(protected\)\|\(var\)\)\s\(static\s\)*\$', ";", s:FOLD_WITH_PHPDOC, 1, 1)
+	" Fold class properties with PhpDoc (public $foo = null;)
+	call s:PHPFoldProperties('^\s*\(static\)\?\s*\(public\|private\|protected\|var\)\s*\(static\)\?\s*\$', ";", s:FOLD_WITH_PHPDOC, 1, 1)
```
Comment (mikehaertl) [inline ftplugin/php/phpfolding.vim:190]: This now also matches any valid order of `static` + visibility:

```
public static $x = 5;
static public $y = 6;
static protected $z = 7;
static var $a = 8;
```

File: ftplugin/php/phpfolding.vim (hunk 1 above)
Comment (mikehaertl) [reply ftplugin/php/phpfolding.vim:184]: Lines matched:

```
function a()
   function b()
public function c()
   public function d()
public        static    function e()
  static    protected function f()
// ...
```

**Review by rayburgemeestre — APPROVED** [review]

**Comment by rayburgemeestre** [comment]
@mikehaertl Thanks again, I uploaded a new version of phpfolding.vim @ vim.org: https://www.vim.org/scripts/script.php?script_id=1623
This should eventually update vim-scripts/phpfolding.vim

**Comment by mikehaertl** [comment]
@rayburgemeestre Cool, thanks for the quick merge and enabling the issues.

**Pull request merged**